src/Compiler/SyntaxTree/WarnScopes.fs

```diff
@@ -0,0 +1,390 @@
+    type private LineMaps = Map<FileIndex, FileIndex * (LineNumber * LineNumber) list>
+
+    type private TempData =
+        {
+            OriginalFileIndex: int
+            mutable WarnDirectives: WarnDirective list
+            mutable LineMaps: LineMaps
+        }
+
+    let private initialData (lexbuf: Lexbuf) =
+        {
+            OriginalFileIndex = lexbuf.StartPos.FileIndex
+            WarnDirectives = []
+            LineMaps = Map.empty
+        }
+
+    let private getTempData (lexbuf: Lexbuf) =
+        lexbuf.GetLocalData("WarnScopeData", (fun () -> initialData lexbuf))
+
+    // *************************************
+    // Collect the line directives during lexing
+    // *************************************
+
+    let RegisterLineDirective (lexbuf, fileIndex, line: int) =
+        let data = getTempData lexbuf
+        let sectionMap = line, lexbuf.StartPos.OriginalLine + 1
+
+        let changer entry =
+            match entry with
+            | None -> Some(data.OriginalFileIndex, [ sectionMap ])
+            | Some(originalFileIndex, maps) ->
+                assert (originalFileIndex = data.OriginalFileIndex) // same lexbuf
+                Some(originalFileIndex, sectionMap :: maps)
+
+        data.LineMaps <- data.LineMaps.Change(fileIndex, changer)
+
+    // *************************************
+    // Collect the warn scopes during lexing
+    // *************************************
+
+    let private getNumber (langVersion: LanguageVersion) m (ns: string) =
+        let argFeature = LanguageFeature.ParsedHashDirectiveArgumentNonQuotes
+
+        let removeQuotes (s: string) =
+            if s.StartsWithOrdinal "\"" && s.EndsWithOrdinal "\"" then
+                if s.StartsWithOrdinal "\"\"\"" && s.EndsWithOrdinal "\"\"\"" then
+                    Some(s.Substring(3, s.Length - 6))
+                else
+                    Some(s.Substring(1, s.Length - 2))
+            elif tryCheckLanguageFeatureAndRecover langVersion argFeature m then
+                Some s
+            else
+                None
+
+        let removePrefix (s: string) =
+            match s.StartsWithOrdinal "FS", langVersion.SupportsFeature argFeature with
+            | true, true -> Some(s.Substring 2, s)
+            | true, false ->
+                warning (Error(FSComp.SR.buildInvalidWarningNumber s, m))
+                None
+            | false, _ -> Some(s, s)
+
+        let parseInt (intString: string, argString) =
+            match System.Int32.TryParse intString with
+            | true, i -> Some i
+            | false, _ ->
+                if langVersion.SupportsFeature argFeature then
+                    warning (Error(FSComp.SR.buildInvalidWarningNumber argString, m))
+
+                None
+
+        ns |> removeQuotes |> Option.bind removePrefix |> Option.bind parseInt
+
+    let private regex =
+        Regex(""" *#(nowarn|warnon|\S+)(?: +([^ \r\n/;]+))*(?:;;)? *(\/\/.*)?$""", RegexOptions.CultureInvariant)
```

Comment:
This code is never called for `#   nowarn` because of the lexer rule that brings us here.
But a description of the regex makes sense.

Reply:
I would consider the fact that the compiler accepts spaces after the hash sign a bug.
But it can certainly be replicated in the new version (it has to happen here and in the lex grammar then).
I guess this should become another RFC discussion.